### PR TITLE
bogovesov homework 3

### PR DESCRIPTION
**Pull request by Bogovesov**
Homework #3.
@internal-jew PTAL

### CODE REVIEW

File: src/main/java/com/alevel/lecture2/BinarySearchAlgorithm.java [hunk 1]
```diff
@@ -13,6 +13,21 @@ static boolean find(int needle, int[] stack) {
...
+            } else {
+                if (needle < stack[mid]) {
+                    high = mid;
+                } else {
+                    low = mid + 1;
+                }
+            }
+        }
         return false;
+
```
Comment (internal-jew) [inline src/main/java/com/alevel/lecture2/BinarySearchAlgorithm.java]: Please, remove redundant empty line.

File: src/main/java/com/alevel/lecture3/MergeSortingAlgorithm.java [hunk 2]
```diff
@@ -7,7 +7,39 @@
 
     @Override
     public int[] sort(int[] source) {
-        // Здесь необходимо реализовать сортировку слиянием.
-        return new int[0];
+        int[] a = mergeSort(source, 0, source.length);
```
Comment (internal-jew) [inline src/main/java/com/alevel/lecture3/MergeSortingAlgorithm.java]: Please, use meaningful variable names. `a` says me nothing.

File: src/main/java/com/alevel/lecture3/MergeSortingAlgorithm.java [hunk 3]
```diff
@@ -7,7 +7,39 @@
...
+        return a;
+    }
+
+    private int[] mergeSort(int[] numbers, int minIndex, int maxIndex) {
+        if (minIndex + 1 < maxIndex) {
+
```
Comment (internal-jew) [inline src/main/java/com/alevel/lecture3/MergeSortingAlgorithm.java]: Redundant empty line.

File: src/main/java/com/alevel/lecture3/MergeSortingAlgorithm.java [hunk 4]
```diff
@@ -7,7 +7,39 @@
...
+            int i = minIndex;
+            int j = midIndex;
+
+            for (int k = 0; k < sizeOfTempArray; k++) {
+
+                boolean condition1 = (j >= maxIndex);
+                boolean condition2 = false;
+
+                if (!condition1)
```
Comment (internal-jew) [inline src/main/java/com/alevel/lecture3/MergeSortingAlgorithm.java]: It is preferred to use `{}` brackets anyway, even if there is only one line of code.

File: src/main/java/com/alevel/lecture3/MergeSortingAlgorithm.java [hunk 5]
```diff
@@ -7,7 +7,39 @@
...
+
+            int sizeOfTempArray = maxIndex - minIndex;
+            int[] tempArray = new int[sizeOfTempArray];
+
+            int i = minIndex;
+            int j = midIndex;
+
+            for (int k = 0; k < sizeOfTempArray; k++) {
+
+                boolean condition1 = (j >= maxIndex);
```
Comment (internal-jew) [inline src/main/java/com/alevel/lecture3/MergeSortingAlgorithm.java]: `condition1` and `condition2` are bad variable names

File: src/main/java/com/alevel/lecture3/QuickSortingAlgorithm.java [hunk 6]
```diff
@@ -9,6 +9,44 @@
...
+        return arrRes;
+    }
+
+    private int[] quickSort(int[] array, int low, int high) {
+
+        if (low >= high)
+            return array;
+
+        int middle = low + (high - low) / 2;
+        int supportingElement = array[middle];
```
Comment (internal-jew) [inline src/main/java/com/alevel/lecture3/QuickSortingAlgorithm.java]: FYI: supporting element is usually called `pivot`.

File: src/main/java/com/alevel/lecture2/BinarySearchAlgorithm.java (hunk 1 above)
Comment (Bogovesov) [reply src/main/java/com/alevel/lecture2/BinarySearchAlgorithm.java]: ок

File: src/main/java/com/alevel/lecture3/MergeSortingAlgorithm.java (hunk 2 above)
Comment (Bogovesov) [reply src/main/java/com/alevel/lecture3/MergeSortingAlgorithm.java]: ok

File: src/main/java/com/alevel/lecture3/MergeSortingAlgorithm.java (hunk 3 above)
Comment (Bogovesov) [reply src/main/java/com/alevel/lecture3/MergeSortingAlgorithm.java]: ok

File: src/main/java/com/alevel/lecture3/MergeSortingAlgorithm.java (hunk 4 above)
Comment (Bogovesov) [reply src/main/java/com/alevel/lecture3/MergeSortingAlgorithm.java]: ok

File: src/main/java/com/alevel/lecture3/MergeSortingAlgorithm.java (hunk 5 above)
Comment (Bogovesov) [reply src/main/java/com/alevel/lecture3/MergeSortingAlgorithm.java]: ok

File: src/main/java/com/alevel/lecture3/QuickSortingAlgorithm.java (hunk 6 above)
Comment (Bogovesov) [reply src/main/java/com/alevel/lecture3/QuickSortingAlgorithm.java]: ok

**Pull request merged**